### ISSUE TITLE
How would this work with Optional Chaining?

### ISSUE THREAD

**Issue by itamark** [rbuckton/proposal-shorthand-improvements#4]
This is a really interesting proposal. 
But I think we should consider this proposal and its implications on this one. https://github.com/tc39/proposal-optional-chaining

I just worry about safety of accessing deeper properties in objects while defining an object like this.

If no optional chaining is present, and something in the chain is undefined, the entire value becomes undefined, but the key remains the last accessor:

```
const a = { o.x.z.y.w }
// o.x.z is undefined
// result: a = { w: undefined }
```
Kind of an implicit optional chaining?

Or, should an error be thrown in this case, because we're accessing `y of undefined`?

**Comment by bergus** [comment]
Yes, evaluation of the property happens as usual and throws an error when the base is `undefined`. The two proposals are independent, but they should work out of the box with each other.
```js
const a = { o.x.z.y.w }; // TypeError
const a = { o?.x?.z?.y?.w }; // {w:undefined} - with the proposal
const a = { ((((o || {}).x || {}).z || {}).y || {}).w }; // {w:undefined} - today
```

**Comment by bergus** [comment]
> they should work out of the box with each other

Oops, I just reviewed the optional-chaining proposal closer, and actually they don't. The problem is that [an *OptionalExpression* such as `o?.x?.z?.y?.w` does immediately evaluate to `undefined`](https://tc39.es/proposal-optional-chaining/#sec-optional-chaining-evaluation) when one of the bases is `null` or `undefined`, instead of evaluating to a *Reference*. The object literal shorthand could not easily derive the target property name then.

And what is worse, they can't even really change that. The proposal suggests a [short-circuiting semantics](https://github.com/tc39/proposal-optional-chaining#short-circuiting) where in the case of
```
const a = { obj?.[getName()] };
```
the `getName()` is never called when `obj` is `null` or `undefined`. We couldn't create a property on `a` since we don't know the key.

The only sensible choice I could imagine was that the shorthand *doesn't create a property* when the expression evaluates to `undefined` - even without optional chains, for consistency:
```
const o = {};
const a = { o.x }; // {}, not {x:undefined}
const a = { o.x.y }; // TypeError
const a = { o?.x }; // {}, not {x:undefined}
const a = { o?.x?.y }; // {}
const a = { o?.x?.[ console.log("doesn't happen"), "y" ] }; // {}
```

**Comment by ByteEater-pl** [comment]
> The only sensible choice I could imagine was that the shorthand _doesn't create a property_ when the expression evaluates to `undefined` - even without optional chains, for consistency

I think it's not very useful, unlike your previous proposal, which seems what programmers would expect.

**Comment by Dragiyski** [comment]
> I think it's not very useful, unlike your previous proposal, which seems what programmers would expect.

I really like the proposal, but I think I would expect optional chaining to not create additional properties. First, the last key might not be deterministic if computed and even if determined there should be a way to optionally assign property. If property is always created then:
```javascript
a = { o?.x?.[getName()]?.test }
```
can be easily rewritten to:
```javascript
a = { test: o?.x?.[getName()]?.test }
```
However, if property must be created conditionally, a semi-equivalent would be
```javascript
{
    let _receiver = o?.x?.[getName()];
    _receiver != null && 'test' in _receiver && o.test = _receiver.test;
}
```
however this is still not equivalent to what chaining does, as the last element is not chained and checked manually. If the ``_receiver`` happens to be ``Proxy`` this would call both ``has`` and ``get`` interceptors, while chained expression would call only ``get``. The only difference in both transformation is the result of ``'test' in o``, but the second case requires several lines of code. Moreover, the first case is still valid if the user requires mandatory property creation.

**Comment by ByteEater-pl** [comment]
> If property is always created then:
> 
> ```js
> a = { o?.x?.[getName()]?.test }
> ```
> 
> can be easily rewritten to:
> 
> ```js
> a = { test: o?.x?.[getName()]?.test }
> ```
OK, that convinced me

**Comment by L3P3** [comment]
A bit off-topic but if we have nothing to insert, we could apply the same rule to `{...x}` so that when x is falsy, it is just not inserting anything. I currently have to write something like this all the time:
```js
{
  ...(x ? x : {}),
}
```
Since that, I expect it to always insert the property. I agree that it is possible by just specifying the key, but even without it, I expect the property to always be set! This is not about what we think should be expected but about what actual people expect, and that was just what I actually expected by reading your code.
BTW optional properties are a pretty nasty thing. JS engines are optimized for specific sets of properties for hidden classes, so when the dev lists properties, they should all be there in a way!

**Comment by ljharb** [comment]
@L3P3 you don’t have to do that, that’s already how it behaves.
